{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "walk-through_consensus_clustering_code_for_students.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "R",
      "name": "ir"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/svafar/sys-bio/blob/main/walk_through_consensus_clustering_code_for_students.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "XMRz4ytrSsOX"
      },
      "source": [
        "This notebook goes over the use of consensus clustering applied to gene expression data from a cohort of acute lymphocytic leukemia (ALL).\n",
        "\n",
        "The code we are running this week is R, not Python. Colab also supports R. To start a new R Colab, just follow this link: [colab.to/r](https://colab.to/r). [RStudio](https://rstudio.com/) is another very convenient tool to develop R code, specially when working on your own."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "RvgupBP8NM8S"
      },
      "source": [
        "# Preliminaries \n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4XT1bJlnuTEB"
      },
      "source": [
        "## 0.1. Install required packages\n",
        "We use Bioconductor, a great tool to install packages, but there are other methods to install libraries, Bioconductor I find it the most convenient."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "5zwEfPyONEok"
      },
      "source": [
        "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
        "  install.packages(\"BiocManager\")\n",
        "\n",
        "BiocManager::install(\"ALL\")  ### expression data\n",
        "BiocManager::install(\"ConsensusClusterPlus\")\n",
        "BiocManager::install(\"cluster\")  ### required for silhouette function\n",
        "BiocManager::install('viridis')  ### nice color palette"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "85C--UJO6nEg"
      },
      "source": [
        "library(ALL)\n",
        "library(ConsensusClusterPlus)\n",
        "library(viridis)\n",
        "library(cluster)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "cBtkjusGqLJk"
      },
      "source": [
        "# 1. Load data\n",
        "We are going to load the ALL data.\n",
        "ALL refers to Acute Lymphoblastic Leukemia and this data set has been generated by the Ritz Laboratory. The original publication is  [*Gene expression profile of adult T-cell acute lymphocytic leukemia identifies distinct subsets of patients with different response to therapy and survival*](https://pubmed.ncbi.nlm.nih.gov/14684422/)\n",
        "\n",
        "The data consist of microarrays from 128 different individuals with acute lymphoblastic leukemia\n",
        "(ALL). A number of additional covariates are available. The data have been normalized (using [RMA](https://ieeexplore.ieee.org/document/6999142))\n",
        "and it is the jointly normalized data that are available here. The data are presented in the form of an\n",
        "exprSet object."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "iLq5SrScJGUo"
      },
      "source": [
        "data(ALL)\n",
        "d = exprs(ALL) # particular syntax to retrieve expression matrix values from e-Set-class"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "3d5DCBf0q1kX"
      },
      "source": [
        "## 1.1. Get familiar with the data"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "BmuvwP1-r8uH"
      },
      "source": [
        "We are working with 128 samples, presumably each from a patient.\n",
        "For each sample, 12,625 genes have been measured."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "q074SQYrq_02"
      },
      "source": [
        "dim(d)\n",
        "head(d)\n",
        "d[1:5,1:5]"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "X5rfQDXBsQtI"
      },
      "source": [
        "Let's inspect the distribution of expression for each patient. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "U1JQGfzVOnRW"
      },
      "source": [
        "boxplot(d[,1:50])"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "q1oVXcp5sbKg"
      },
      "source": [
        "Remember, we are plotting the gene expression distributions per sample, ie, patient. That means that each boxplot has ~12,000 values. As you can see, expression values are very nicely distributed between 2 and 14 with some extreme values. From this distribution, we can conclude that the data we have in hands comes from microarray data and not from RNA-seq."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "M7b4lfdrvTCy"
      },
      "source": [
        "# Filter data"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ssNPWZM4teC7"
      },
      "source": [
        "Let's filter out non-variable genes"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "klIdXLpgbwG3"
      },
      "source": [
        "mads=apply(d,1,mad)\n",
        "d=d[rev(order(mads))[1:5000],]\n",
        "d[1:5,1:5]"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "3s1dcv4ctBbB"
      },
      "source": [
        "Still, there is some variation between patients so we will scale the data to their median expression, so they are even more comparable. Other apporaches like z-score scaling are other valid options."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "oi5D3W1mb1jf"
      },
      "source": [
        "d = sweep(d, 1, apply(d,1,median,na.rm=T)) # sweep removes, in this case substracts, the value you define, in this case the median"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "wSaEe7jkVfl9"
      },
      "source": [
        "d[1:5,1:5]"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "bgdiHbFdtyxx"
      },
      "source": [
        "# 3. Consensus clustering"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Xj5knL-7b-lX"
      },
      "source": [
        "maxK2test = 8\n",
        "clusters = ConsensusClusterPlus(d, \n",
        "  maxK=maxK2test, # max number of K to test\n",
        "  reps=50,  # number of subsample iterations\n",
        "  pItem=0.8, # proportion of items (patients) to sample\n",
        "  pFeature=1, # proportion of features (genes) to subsample\n",
        "  clusterAlg=\"hc\", # clustering algorithm. \n",
        "  distance=\"pearson\", # distance between samples\n",
        "  innerLinkage=\"complete\", # hierarchical linkage method\n",
        "  seed=1262118388.71279) # optional numerical value to control reproducibility"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "7ayXJPEZWh7E"
      },
      "source": [
        "# 4. Measures to identify optimal *K*\n",
        "All these measures are heuristic methods and can be difficult to unambiguously identify the number of clusters. Interpretation by the scientist is key for choosing a relevant *K*. \n",
        "\n",
        "## 4.1. Elbow plot.\n",
        "When using [this method](https://en.wikipedia.org/wiki/Elbow_method_(clustering)) look for saturation on the amount of variance explained. The more *K*, the more variance will be explained, but at the cost of having more clusters (over-fitting).  \n",
        "\n",
        "## 4.2. Proportion of ambiguous clustering (PAC)\n",
        "Consensus index refer to the probability of two samples clustering in the same cluster. Therefore the top-right indicates elements that always cluster together. On the contrary, the bottom left are elements that never cluster together. Therefore, the middle section quantifies the elements whose cluster assignment is most ambiguous. Our objective is to limit that, so the lower the increase in that area, the better. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "4B6PzSgyPIwR"
      },
      "source": [
        "############## PAC implementation ##############\n",
        "Kvec = 2:maxK2test\n",
        "x1 = 0.1; x2 = 0.9 # threshold defining the intermediate sub-interval\n",
        "PAC = rep(NA,length(Kvec)) \n",
        "names(PAC) = paste(\"K=\",Kvec,sep=\"\") # from 2 to maxK\n",
        "\n",
        "for(i in Kvec){\n",
        "  M = clusters[[i]]$consensusMatrix\n",
        "  Fn = ecdf(M[lower.tri(M)])\n",
        "  PAC[i-1] = Fn(x2) - Fn(x1)\n",
        "  print(c(i, Fn(x2) - Fn(x1)))\n",
        "}#end for i\n",
        "\n",
        "# The optimal K\n",
        "optK = Kvec[which.min(PAC)]\n",
        "print(optK)\n",
        "########################################################"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Dh-gqEvVdei-"
      },
      "source": [
        "# 4.3. Silhouette method\n",
        "The silhouette value is a measure of how similar an object is to its own cluster (cohesion) compared to other clusters (separation). The silhouette ranges from −1 to +1, where a high value indicates that the object is well matched to its own cluster and poorly matched to neighboring clusters. If most objects have a high value, then the clustering configuration is appropriate. If many points have a low or negative value, then the clustering configuration may have too many or too few clusters."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "NSMQFKbHFXv3"
      },
      "source": [
        "lsil = vector(\"list\",(maxK2test-1))\n",
        "\n",
        "for(i in 2:maxK2test){\n",
        "  sil = silhouette(clusters[[i]]$consensusClass, dist(t(d), method = \"euclidean\"))\n",
        "  sizes = table(clusters[[i]]$consensusClass)\n",
        "  plot(sil, col=rep(clusters[[i]]$clrs[[3]], rep=sizes), main=paste(\"K=\",i))\n",
        "  \n",
        "  sil_average = mean(sil[, 3])\n",
        "  abline(v=sil_average)\n",
        "  lsil[[i-1]]=sil\n",
        "\n",
        "  print(sizes)\n",
        "  print(rep(clusters[[i]]$clrs[[3]])) # this is simply to define a different color per cluster\n",
        " } "
      ],
      "execution_count": null,
      "outputs": []
    }
  ]
}